pkg/registry/storage/volumeattachment/strategy.go

```diff
@@ -98,15 +106,22 @@ func (volumeAttachmentStrategy) PrepareForUpdate(ctx context.Context, obj, old r
 	if requestInfo, found := genericapirequest.RequestInfoFrom(ctx); found {
 		groupVersion = schema.GroupVersion{Group: requestInfo.APIGroup, Version: requestInfo.APIVersion}
 	}
+
+	newVolumeAttachment := obj.(*storage.VolumeAttachment)
+	oldVolumeAttachment := old.(*storage.VolumeAttachment)
+
 	switch groupVersion {
 	case storageapiv1beta1.SchemeGroupVersion:
 		// allow modification of Status via main resource for v1beta1
 	default:
-		newVolumeAttachment := obj.(*storage.VolumeAttachment)
-		oldVolumeAttachment := old.(*storage.VolumeAttachment)
 		newVolumeAttachment.Status = oldVolumeAttachment.Status
 		// No need to increment Generation because we don't allow updates to spec
 	}
+
+	// disable InlineVolumeSource
+	if !utilfeature.DefaultFeatureGate.Enabled(features.CSIInlineVolume) && !inlineCSIInUse(oldVolumeAttachment.Spec.Source.InlineVolumeSource) {
+		newVolumeAttachment.Spec.Source.InlineVolumeSource = nil
```

Comment:
this is confusing, so what happens if feature is not enabled and inlineCSIInuse ? then we dont mutate ?

Reply:
This is to handle the scenario where a user downgrades to a version where the feature is off.  The feature will cease to function, but we allow the object to be deleted from the API server.

Before, the pattern was to check for the feature gate in validation. If a user downgraded, the feature would stop working, but they could not modify or delete the API object.